clusterversion: bump minBinary to 22.2

celiala opened the pull request:
This PR bumps binaryMinSupportedVersion to 22.2 before cutting the 23.1
release branch, as part of the stability period process.

In previous stability periods, the work of deleting pre-22.2 gates/tests
and bumping binaryMinSupportedVersion was done at the same time. For 23.1,
we will break this up into smaller tasks:

- This PR will just bump binaryMinSupportedVersion to 22.2, skipping 22.2
  mixed-version tests as needed to bump this value.
- The clean-up of pre-22.2 gates and 22.2 mixed-version tests will be done
  via follow-up work/issues.

Issue: #95530
Epic: CRDB-23572

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/96656)
<!-- Reviewable:end -->

RaduBerinde commented:
The version upgrade test fails with:
```
failed to create engines: store last used with cockroach version v22.2 is too old for running version v1000022.2-36 (which requires data from v1000022.2 or later)
```

This is caused by the 1000000 `DevOffset`. When we want to allow upgrades to dev (which is needed in upgrade tests), all keys other than the first non-primordial get offset. We want the min-binary-version = v22_2 key to not be offset, so it needs to be the first (non-primordial) key.

Here is what I think is needed:
 - remove all code that uses `clusterversion.TODOPreV22_1`
 - remove all keys before V22_2 (i.e. `V22_1` and all `V22_2foo`)
 - add a `clusterversion.TODOPreV22_2` and replace all occurrences of removed keys with this

Maybe @dt or @andreimatei can double check that this makes sense.

RaduBerinde commented:
I opened [#96701](https://github.com/cockroachdb/cockroach/pull/96701) for the `clusterversion.TODOPreV22_1` removal.

RaduBerinde commented:
We should also update the `DevOffset` -ing code to either not offset anything up to `BinaryMinSupportedVersionKey`, or assert that the first non-primordial key is `BinaryMinSupportedVersionKey`.

RaduBerinde commented:
> We should also update the DevOffset -ing code to either not offset anything up to BinaryMinSupportedVersionKey,

Opened https://github.com/cockroachdb/cockroach/pull/96702 for this. With that, we won't need to remove the older keys in the same change (thanks @dt for the suggestion!).

RaduBerinde commented:
@celiala #96702 merged, I *think* (and hope) the cluster version failure will go away if you rebase.

celiala committed:
clusterversion: bump minBinary to 22.2

This PR bumps binaryMinSupportedVersion to 22.2 before cutting the 23.1
release branch, as part of the stability period process.

In previous stability periods, the work of deleting pre-22.2 gates/tests
and bumping binaryMinSupportedVersion was done at the same time. For 23.1,
we will break this up into smaller tasks:

- This PR will just bump binaryMinSupportedVersion to 22.2, skipping 22.2
  mixed-version tests as needed to bump this value.
- The clean-up of pre-22.2 gates and 22.2 mixed-version tests will be done
  via follow-up work/issues.

Issue: #95530
Epic: CRDB-23572

Release note: None

celiala commented:
> @celiala https://github.com/cockroachdb/cockroach/pull/96702 merged, I think (and hope) the version upgrade failure will go away if you rebase.

Thank you for all the help unblocking this! Okay, I've rebased and will (hopefully) pass

RaduBerinde commented:
acceptance/version-upgrade passed :)

Looks like we need to skip `BenchmarkTimeBoundIterate` for now.

celiala commented:
The tests passed 🎉 -- thanks again for fixing @RaduBerinde!

This is RFAL!

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96656#-:-NNheqFjCkglBT7w8QNY:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt and @itsbilal)



<!-- Sent from Reviewable.io -->

celiala commented:
TFTR!

bors r=RaduBerinde
